### PR TITLE
Add SpiderMonkey backward compatibility

### PR DESCRIPTION
**Pull request by jiahuili430**
## Overview

- Use `for in` to fix syntax error when using SpiderMonkey 1.8.5

```
[2024-12-12T22:46:32.210Z] Failed to compile script.
[2024-12-12T22:46:32.210Z] SyntaxError: missing ; after for-loop initializer
[2024-12-12T22:46:32.210Z] Stacktrace:
[2024-12-12T22:46:32.210Z] 	@:0
```

- Add backward compatibility

Related PR: https://github.com/apache/couchdb/pull/5321

## Checklist

- [ ] Code is written and works correctly
- [ ] Changes are covered by tests
- [ ] Any new configurable parameters are documented in `rel/overlay/etc/default.ini`
- [ ] Documentation changes were made in the `src/docs` folder
- [ ] Documentation changes were backported (separated PR) to affected branches

### CODE REVIEW

**Comment by jiahuili430** [comment]
Tested in the [`jenkins-sm-backward-compatibility`](https://github.com/apache/couchdb/compare/main...jenkins-sm-backward-compatibility) branch.
image

**Comment by nickva** [comment]
Oops, it does look like we broke SM 1.8.5 support. It's still one of the fastest SM engine versions out of the box, especially when using custom reduces.

**Review by nickva — APPROVED** [review]

File: share/server/util.js [hunk 1]
```diff
@@ -116,7 +116,14 @@ var Couch = {
         "Expression does not eval to a function. (" + source.toString() + ")"]);
     };
   },
-  recursivelySeal : deepFreeze
+  recursivelySeal: function (obj) {
```
Comment (janl) [inline share/server/util.js:119]: this effectively bypasses the new deepFreeze() function we introduced for this. Can’t we do duck-typing here and call deepFreeze if it is available and seal if it isn’t?

File: share/server/util.js [hunk 2]
```diff
@@ -116,7 +116,14 @@ var Couch = {
...
+    seal(obj);
+    for (var propname in obj) {
```
Comment (janl) [inline share/server/util.js:121]: What does this for-loop do?

Comment (jiahuili430) [reply share/server/util.js:121]: `propname` is the index, if the element type is `object`, this `function(obj) { seal(obj); ... }` will be called again with the new argument `obj = obj[propname]`.

File: share/server/util.js (hunk 1 above)
Comment (jiahuili430) [reply share/server/util.js:119]: Tried the following 2 methods but get same errors in `chttpd_view_test` when using `SpiderMonkey 1.8.5`.

```diff
recursivelySeal: function (obj) {
-    seal(obj);
+    deepFreeze(obj);
    for (var propname in obj) {
      if (typeof obj[propname] == "object") {
        arguments.callee(obj[propname]);
      }
    }
  },
```

```diff
+  recursivelySeal: deepFreeze
};
...
function deepFreeze(object) {
...
    // Freeze properties before freezing self
    for (var i in propNames) {
        const value = object[propNames[i]];

-      if ((value && typeof value === "object") || typeof value === "function") {
+      if (value && typeof value == "object") {
            deepFreeze(value);
        }
    }
}
```

Error message:
```
[2024-12-23T15:18:42.050Z] module 'chttpd_view_test'
[2024-12-23T15:18:42.050Z]   chttpd view tests
[2024-12-23T15:18:45.302Z]     chttpd_view_test:74: -view_test_/0-fun-30- (t_view_with_queries_keys)...*failed*
[2024-12-23T15:18:45.302Z] in function chttpd_view_test:t_view_with_queries_keys/1 (test/eunit/chttpd_view_test.erl, line 132)
[2024-12-23T15:18:45.302Z] in call from eunit_test:run_testfun/1 (eunit_test.erl, line 71)
[2024-12-23T15:18:45.302Z] in call from eunit_proc:run_test/1 (eunit_proc.erl, line 543)
[2024-12-23T15:18:45.302Z] in call from eunit_proc:with_timeout/3 (eunit_proc.erl, line 368)
[2024-12-23T15:18:45.302Z] in call from eunit_proc:handle_test/2 (eunit_proc.erl, line 526)
[2024-12-23T15:18:45.302Z] in call from eunit_proc:tests_inorder/3 (eunit_proc.erl, line 468)
[2024-12-23T15:18:45.302Z] in call from eunit_proc:with_timeout/3 (eunit_proc.erl, line 358)
[2024-12-23T15:18:45.302Z] in call from eunit_proc:run_group/2 (eunit_proc.erl, line 582)
[2024-12-23T15:18:45.302Z] **error:{assertMatch,[{module,chttpd_view_test},
[2024-12-23T15:18:45.302Z]               {line,132},
[2024-12-23T15:18:45.302Z]               {expression,"Res"},
[2024-12-23T15:18:45.302Z]               {pattern,"# { << \"results\" >> := [ # { << \"total_rows\" >> := 3 , << \"offset\" >> := 1 , << \"rows\" >> := [ # { << \"id\" >> := << \"a\" >> } , # { << \"id\" >> := << \"c\" >> } ] } ] }"},
[2024-12-23T15:18:45.302Z]               {value,#{<<"error">> => <<"ReferenceError">>,
[2024-12-23T15:18:45.302Z]                        <<"reason">> =>
[2024-12-23T15:18:45.302Z]                            <<"{[{<<\"message\">>,<<\"Reflect is not d"...>>}}]}
[2024-12-23T15:18:45.302Z]   output:<<"">>
```

**Pull request merged**